python/beeai_framework/context.py

```diff
@@ -45,56 +44,52 @@ class RunContextInput(BaseModel):
     signal: AbortSignal | None = None
 
 
-class Run:
-    def __init__(self, handler: Callable[[], R], context: GetRunContext) -> None:
+class Run(Generic[R]):
+    def __init__(self, handler: Callable[[], R | Awaitable[R]], context: "RunContext") -> None:
         super().__init__()
-        self.handler = handler
+        self.handler = ensure_async(handler)
         self.tasks: list[tuple[Callable[[Any], None], Any]] = []
         self.run_context = context
 
-    def __await__(self) -> R:
+    def __await__(self) -> Generator[Any, None, R]:
         return self._run_tasks().__await__()
 
     def observe(self, fn: Callable[[Emitter], Any]) -> Self:
         self.tasks.append((fn, self.run_context.emitter))
         return self
 
-    def context(self, context: GetRunContext) -> Self:
+    def context(self, context: "RunContext") -> Self:
         self.tasks.append((self._set_context, context))
         return self
 
-    def middleware(self, fn: Callable[[GetRunContext], None]) -> Self:
+    def middleware(self, fn: Callable[["RunContext"], None]) -> Self:
         self.tasks.append((fn, self.run_context))
         return self
 
     async def _run_tasks(self) -> R:
         for fn, param in self.tasks:
-            if inspect.iscoroutinefunction(fn):
-                await fn(param)
-            else:
-                fn(param)
+            await ensure_async(fn)(param)
+
         self.tasks.clear()
         return await self.handler()
 
-    def _set_context(self, context: GetRunContext) -> None:
+    def _set_context(self, context: "RunContext") -> None:
         self.run_context.context = context
         self.run_context.emitter.context = context
 
 
 class RunContext(RunInstance):
     storage: ContextVar[Self] = ContextVar("storage", default=None)
 
-    def __init__(
-        self, *, instance: RunInstance, context_input: RunContextInput, parent: GetRunContext | None = None
-    ) -> None:
+    def __init__(self, *, instance: RunInstance, context_input: RunContextInput, parent: Self | None = None) -> None:
```

Comment:
ty for finding the solution to get rid of `GetRunContext` it kept eluding me when I tried

Reply:
same story here, did my best to get rid of it